src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,137 @@
+            increaseElementDataLength();
+            addValueToElementData(value);
+        } else {
+            rightShiftOfElements(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] bufferArray = (T[]) new Object[list.size()];
+        for (int i = 0; i < list.size(); i++) {
+            bufferArray[i] = list.get(i);
+        }
+        if (elementData.length - size < list.size()) {
+            increaseElementDataLength();
+        }
+        System.arraycopy(bufferArray, INITIAL_INDEX, elementData, size, bufferArray.length);
+        size += bufferArray.length;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        } else {
+            return elementData[index];
+        }
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        } else {
+            elementData[index] = value;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < INITIAL_INDEX) {
+            throw new ArrayListIndexOutOfBoundsException(TEXT_OF_EXCEPTION);
+        }
+        if (index == size - 1) {
+            size--;
+            return elementData[index];
+        } else {
+            T removeValue = elementData[index];
+            leftShiftOfElements(index);
+            return removeValue;
+        }
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == null) {
+                if (null == elementData[i]) {
+                    leftShiftOfElements(i);
+                    return null;
+                }
+            } else if (element.equals(elementData[i])) {
+                leftShiftOfElements(i);
+                return element;
+            }
+        }
+        throw new NoSuchElementException();
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == INITIAL_INDEX;
+    }
+
+    private void increaseElementDataLength() {
+        elementData = Arrays.copyOf(elementData, (int) (elementData.length * 1.5 + 1));
+    }
+
+    private void rightShiftOfElements(T value, int index) {
+        T[] bufferElementData = (T[]) new Object[elementData.length + 1];
+        System.arraycopy(elementData, INITIAL_INDEX, bufferElementData, INITIAL_INDEX, index);
+        System.arraycopy(elementData, index, bufferElementData, index + 1, size);
+        bufferElementData[index] = value;
+        elementData = bufferElementData;
+        size++;
+    }
+
+    private void addValueToElementData(T value) {
```

Comment:
probably better to no create this method because it is not only add value but increase the size. and when we just read the code in place where method is called it can be not obvious.